client/network/src/service/out_events.rs

```diff
@@ -0,0 +1,236 @@
+	stats: Arc<Mutex<Option<Arc<Mutex<StatsCollect>>>>>,
+}
+
+impl Stream for Receiver {
+	type Item = Event;
+
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Event>> {
+		if let Some(ev) = ready!(Pin::new(&mut self.inner).poll_next(cx)) {
+			let stats = self.stats.lock();
+			if let Some(stats) = stats.as_ref() {
+				decrease_stats(&mut *stats.lock(), &ev);
+			} else {
+				log::warn!("Inconsistency in out_events: event happened before sender associated");
+			}
+			Poll::Ready(Some(ev))
+		} else {
+			Poll::Ready(None)
+		}
+	}
+}
+
+impl fmt::Debug for Receiver {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		f.debug_tuple("Receiver").finish()
+	}
+}
+
+impl Drop for Receiver {
+	fn drop(&mut self) {
+		// Empty the list to properly decrease the stats.
+		while let Some(Some(_)) = self.next().now_or_never() {}
+	}
+}
+
+/// Statistics about the content of the channel.
+#[non_exhaustive]
+pub struct StatsCollect {
+	/// For each protocol, total number of substream open notifications in the channel.
+	pub notifications_open_messages_count: HashMap<ConsensusEngineId, u64>,
+
+	/// For each protocol, total number of substream closed notifications in the channel.
+	pub notifications_closed_messages_count: HashMap<ConsensusEngineId, u64>,
+
+	/// For each protocol, the total size of all notifications stuck in the channel.
+	pub notifications_sizes_total: HashMap<ConsensusEngineId, u64>,
+
+	/// Number of DHT events in the channel.
+	///
+	/// > **Note**: DHT events are extremely rare at the moment, but this could be broken down
+	/// >			further if there's a need to.
+	pub dht_events: u64,
+}
+
+/// Collection of senders.
+pub struct OutChannels {
+	event_streams: Vec<Sender>,
+	/// The stats we collect. A clone of this is sent to each [`Receiver`] associated with this
+	/// object.
+	stats: Arc<Mutex<StatsCollect>>,
+}
+
+impl OutChannels {
+	/// Creates a new empty collection of senders.
+	pub fn new() -> Self {
+		OutChannels {
+			event_streams: Vec::new(),
+			stats: Arc::new(Mutex::new(StatsCollect {
+				notifications_open_messages_count: HashMap::new(),
+				notifications_closed_messages_count: HashMap::new(),
+				notifications_sizes_total: HashMap::new(),
+				dht_events: 0,
+			}))
+		}
+	}
+
+	/// Adds a new [`Sender`] to the collection.
+	pub fn push(&mut self, sender: Sender) {
+		let mut stats = sender.stats.lock();
+		debug_assert!(stats.is_none());
+		*stats = Some(self.stats.clone());
+		drop(stats);
+		self.event_streams.push(sender);
+	}
+
+	/// Returns the number of senders in the collection.
+	pub fn len(&self) -> usize {
+		self.event_streams.len()
+	}
+
+	/// Sends an event.
+	pub fn send(&mut self, event: Event) {
+		self.event_streams.retain(|sender| {
+			sender.inner.unbounded_send(event.clone()).is_ok()
+		});
+
+		// The number of elements remaining in `event_streams` corresponds to the number of times
+		// we have sent an element on the channel.
```

Comment:
```suggestion
		// we have sent an element on the channels.
```

Reply:
I am not sure the comment is still up to date. If so, I would have expected it to be on `increase_metrics` itself.